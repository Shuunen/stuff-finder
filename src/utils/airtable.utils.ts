/* eslint-disable functional/immutable-data */
import { safeParse } from 'valibot'
import { emptyItem } from '../constants'
import { del, get, patch, post } from './browser.utils'
import { logger } from './logger.utils'
import { airtableDeleteResponseSchema, airtableMultipleResponseSchema, airtableSingleResponseSchema, type AirtableSingleRecordResponse, type Item, type ItemField, type ItemPhoto } from './parsers.utils'
import { state } from './state.utils'

const airtableBaseUrl = 'https://api.airtable.com/v0'

export const airtableMaxRequestPerSecond = 5

// biome-ignore lint/complexity/noExcessiveCognitiveComplexity: <explanation>
export function getItemFieldsToPush (data: Item, currentState = state) { // eslint-disable-line max-statements, complexity, sonarjs/cognitive-complexity
  const fields: Partial<AirtableSingleRecordResponse['fields']> = {}
  if (data.barcode.length > 0) fields.barcode = data.barcode
  if (data.box.length > 0) fields.box = data.box
  if (data.brand.length > 0) fields.brand = data.brand
  if (data.category.length > 0) fields.category = data.category
  if (data.details.length > 0) fields.details = data.details
  if (data.drawer.length > 0) fields.drawer = data.drawer
  if (data.location.length > 0) fields.location = data.location
  if (data.name.length > 0) fields.name = data.name
  /* c8 ignore next 2 */
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  if (data.photo !== undefined && data.photo.length > 0 && fields.photo?.[0]?.url !== data.photo[0]?.url) fields.photo = [{ url: data.photo[0]?.url } as unknown as ItemPhoto] // we don't need the whole object
  // eslint-disable-next-line @typescript-eslint/no-magic-numbers
  if (data.price !== undefined && data.price > -1) fields.price = data.price
  if (data.reference.length > 0) fields.reference = data.reference
  if (data.status.length > 0) fields.status = data.status
  fields['ref-printed'] = data['ref-printed']
  if (data.id.length > 0) {
    const existing = currentState.items.find(existingItem => existingItem.id === data.id)
    if (!existing) throw new Error('existing item not found locally')
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    const dataFields = Object.keys(fields) as ItemField[]
    for (const field of dataFields) {
      /* c8 ignore next 2 */
      if (field === 'id') continue // eslint-disable-line no-continue
      const hasSamePhoto = (field === 'photo' && existing.photo && existing.photo[0]?.url === fields.photo?.[0]?.url) ?? false
      const hasSameValue = existing[field] === fields[field]
      // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
      if (hasSamePhoto || hasSameValue) delete fields[field]
    }
  }
  return fields
}

export function airtableRecordToItem (record: AirtableSingleRecordResponse) {
  return {
    ...emptyItem,
    ...record.fields,
    id: record.id,
  }
}

export function isLocalAndRemoteSync (records: AirtableSingleRecordResponse[], currentState = state) {
  const record = records.at(0)
  if (!record) throw new Error('remoteFirst is undefined')
  const remoteFirst = airtableRecordToItem(record)
  const localFirst = currentState.items.at(0)
  const isFirst = localFirst === undefined ? false : remoteFirst.id === localFirst.id && remoteFirst['updated-on'] === localFirst['updated-on']
  if (isFirst) return true
  const localLast = currentState.items.at(-1) // eslint-disable-line @typescript-eslint/no-magic-numbers
  return localLast === undefined ? false : remoteFirst.id === localLast.id && remoteFirst['updated-on'] === localLast['updated-on']
}

export async function getOneItem (id: Item['id'], getMethod = get) {
  const { base, table } = state.credentials
  const url = `${airtableBaseUrl}/${base}/${table}/${id}`
  const result = safeParse(airtableSingleResponseSchema, await getMethod(url))
  if (!result.success) {
    logger.error(result.issues)
    throw new Error(`failed to fetch item, issue(s) : ${result.issues.map(issue => issue.message).join(', ')}`)
  }
  return airtableRecordToItem(result.output)
}

export async function getAllItems (offset?: string, getMethod = get) {
  const offsetParameter = offset === undefined ? '' : `&offset=${offset}`
  const sortByUpdatedFirst = '&sort%5B0%5D%5Bfield%5D=updated-on&sort%5B0%5D%5Bdirection%5D=desc'
  const { base, table, view } = state.credentials
  const url = `${airtableBaseUrl}/${base}/${table}?view=${view}${offsetParameter}${sortByUpdatedFirst}`
  const result = safeParse(airtableMultipleResponseSchema, await getMethod(url))
  if (!result.success) {
    logger.error(result.issues)
    throw new Error(`failed to fetch item, issue(s) : ${result.issues.map(issue => issue.message).join(', ')}`)
  }
  return result.output
}

export async function deleteItemRemotely (id: Item['id'], currentState = state, delMethod = del) {
  const { base, table } = currentState.credentials
  const url = `${airtableBaseUrl}/${base}/${table}/${id}`
  return safeParse(airtableDeleteResponseSchema, await delMethod(url))
}

// eslint-disable-next-line @typescript-eslint/max-params
export async function pushItemRemotely (item: Item, currentState = state, postMethod = post, patchMethod = patch) {
  const fields = getItemFieldsToPush(item, currentState)
  if (Object.keys(fields).length === 0) {
    logger.showLog('no update to push')
    return { output: { fields, id: item.id }, success: false } // eslint-disable-line @typescript-eslint/naming-convention
  }
  const data = { fields }
  const { base, table } = currentState.credentials
  const baseUrl = `${airtableBaseUrl}/${base}/${table}`
  if (item.id === '') return safeParse(airtableSingleResponseSchema, await postMethod(baseUrl, data))
  const url = `${baseUrl}/${item.id}`
  return safeParse(airtableSingleResponseSchema, await patchMethod(url, data))
}
