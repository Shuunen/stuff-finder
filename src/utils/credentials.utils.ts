import type { AppCredentials } from '../types'

export function parseClipboard (clipboard: string) {
  // clipboard should contains something like : "appABC
  // patXYZ.123
  // my-table
  // my-view
  // azerty4567"
  const regex = /"(?<base>app\w+)\n(?<token>pat[\w.]+)\n(?<table>[\w-]+)\n(?<view>[\w-]+)\n(?<wrap>[\w-]+)"/u
  const { base = '', table = '', token = '', view = '', wrap = '' } = regex.exec(clipboard)?.groups ?? {}
  return { base, table, token, view, wrap } satisfies AppCredentials
}
