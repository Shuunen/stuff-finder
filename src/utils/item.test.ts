
import { Result, clone, nbMsInMinute, sleep } from 'shuutils'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import { databaseMock } from './database.mock'
import { removeAppWriteFields } from './database.utils'
import { addItem, areItemsEquivalent, boxStringToBox, deleteItem, drawerStringToDrawer, formToItem, getItems, isDataOlderThan, itemBoxToRoom, itemForm, itemToForm, itemToLocation, statusStringToStatus, updateItem } from './item.utils'
import { logger } from './logger.utils'
import { mockFetch, mockItem, mockState } from './mock.utils'

vi.mock('appwrite', () => databaseMock.appwrite)

globalThis.fetch = mockFetch

describe('item.utils', () => {

  beforeEach(() => {
    mockFetch.mockClear()
    databaseMock.reset()
  })

  it('mockItem A', () => {
    const name = 'super glass tempered shovel'
    const item = mockItem({ name })
    expect(item.name).toBe(name)
    expect(item.photos).toHaveLength(2)
  })

  it('addItem A should create remotely & push locally', async () => {
    const item = mockItem({ $id: '' })
    const state = mockState({ items: [mockItem({ $id: '1' }), mockItem({ $id: '2' })] })
    expect(state.items).toHaveLength(2)
    const result = await addItem(item, state)
    expect(result.ok).toBe(true) // should be a success to have added the item remotely & locally
    expect(result).toMatchSnapshot()
    expect(databaseMock.createDocument).toHaveBeenNthCalledWith(1, 'databaseA', 'collectionA', "reference-b", removeAppWriteFields(item))
    expect(state.items).toHaveLength(3) // because we just added a new item, the number of items should increase
  })

  it('updateItem A should update remotely & update locally', async () => {
    const item1 = mockItem({ $id: '1' })
    const item2 = mockItem({ $id: '2' })
    const state = mockState({ items: [item1, item2] })
    expect(state.items).toHaveLength(2)
    const result = await updateItem(item1, state)
    expect(result.ok).toBe(true) // should be a success to have updated the item remotely & locally
    expect(result).toMatchSnapshot()
    expect(databaseMock.updateDocument).toHaveBeenNthCalledWith(1, 'databaseA', 'collectionA', '1', removeAppWriteFields(item1))
    expect(state.items).toHaveLength(2) // because we just updated an existing item, the number of items should not change
  })

  it('formToItem A default itemForm', () => {
    const item = formToItem(itemForm)
    expect(item).toMatchSnapshot()
    expect(item.photos).toHaveLength(0)
  })

  it('formToItem B itemForm with some values', () => {
    const form = clone(itemForm)
    const url = 'https://photos.org/200/200'
    form.fields.price.value = '42'
    form.fields.photo.value = url
    const item = formToItem(form)
    expect(item).toMatchSnapshot()
    expect(item.price).toBe(42)
    expect(item.photos).toHaveLength(1)
    expect(item.photos[0]).toMatchInlineSnapshot(`"https://photos.org/200/200"`)
  })

  it('deleteItem A delete an item in state', async () => {
    logger.disable()
    const itemA = mockItem({ $id: 'itemA' })
    const stateA = mockState({ items: [itemA, mockItem({ $id: 'itemB' })] })
    const result = await deleteItem(itemA, stateA)
    expect(result.ok).toBe(true)
    expect(databaseMock.deleteDocument).toHaveBeenNthCalledWith(1, 'databaseA', 'collectionA', 'itemA')
    expect(JSON.stringify(Result.unwrap(result).value)).toMatchInlineSnapshot(`"{"$id":"itemA","collectionId":"collectionA","databaseId":"databaseA","isThisMockedDataFromMock":true}"`)
    logger.enable()
  })

  it('deleteItem B delete an item not in state', async () => {
    logger.disable()
    const itemA = mockItem({ $id: 'itemA' })
    const itemB = mockItem({ $id: 'non-existing-id' })
    const stateA = mockState({ items: [itemA] })
    expect(stateA.items).toHaveLength(1)
    const result = await deleteItem(itemB, stateA)
    const { error } = Result.unwrap(result)
    expect(result.ok).toBe(false)
    expect(error).toMatchInlineSnapshot(`"item not found in state"`)
    expect(stateA.items).toHaveLength(1)
    expect(databaseMock.deleteDocument).toHaveBeenNthCalledWith(1, 'databaseA', 'collectionA', 'non-existing-id')
    logger.enable()
  })

  it('areItemsEquivalent A should be equivalent', () => {
    const itemLeft = mockItem({ $id: 'same' })
    const itemRight = mockItem({ $id: 'same' })
    expect(areItemsEquivalent(itemLeft, itemRight)).toBe(true)
  })

  it('areItemsEquivalent B should be equivalent because we dont check id', () => {
    const itemLeft = mockItem({ $id: 'same' })
    const itemRight = mockItem({ $id: 'different' })
    expect(areItemsEquivalent(itemLeft, itemRight)).toBe(true)
  })

  it('areItemsEquivalent C should not be equivalent', () => {
    const itemLeft = mockItem({ $id: 'same' })
    const itemRight = mockItem({ $id: 'different', name: 'something else' })
    expect(areItemsEquivalent(itemLeft, itemRight)).toBe(false)
  })

  it('areItemsEquivalent D should not be equivalent because of photo', () => {
    const itemLeft = mockItem({ $id: 'same' })
    const itemRight = mockItem({ $id: 'same', photos: [] })
    expect(areItemsEquivalent(itemLeft, itemRight)).toBe(false)
  })

  it('itemToForm A mocked item', () => {
    const item = mockItem({ $id: 'itemToForm-A' })
    const form = itemToForm(item)
    expect(form).toMatchSnapshot()
  })

  it('itemToForm B mocked item without photo', () => {
    const item = mockItem({ $id: 'itemToForm-B', photos: [] })
    const form = itemToForm(item)
    expect(form.fields.photo.value).toBe('')
  })

  it('itemToForm C no item', () => {
    const form = itemToForm()
    expect(form).toMatchSnapshot()
  })

  it('itemToForm D no price', () => {
    const item = mockItem({ price: -1 })
    const form = itemToForm(item)
    expect(form.fields.price.value).toBe('')
  })

  it('itemToForm E with price', () => {
    const item = mockItem({ price: 42 })
    const form = itemToForm(item)
    expect(form.fields.price.value).toBe('42')
  })

  it('itemToLocation A complete data', () => {
    expect(itemToLocation(mockItem())).toMatchInlineSnapshot(`"Entrée B‧2 (usb & audio)"`)
  })

  it('itemToLocation B no box', () => {
    const item = mockItem({ box: '' })
    expect(itemToLocation(item)).toMatchInlineSnapshot(`""`)
  })

  it('itemToLocation C partial box without details', () => {
    // @ts-expect-error for testing purposes
    const item = mockItem({ box: 'G' })
    expect(itemToLocation(item)).toMatchInlineSnapshot(`"Salon G‧2"`)
  })

  it('itemToLocation D no drawer', () => {
    // @ts-expect-error for testing purposes
    const item = mockItem({ box: 'zboubi' })
    item.drawer = -1
    expect(itemToLocation(item)).toMatchInlineSnapshot(`"Z"`)
  })

  it('itemBoxToRoom A', () => { expect(itemBoxToRoom('A (apple)')).toMatchInlineSnapshot(`"entrée"`) })
  // @ts-expect-error for testing purposes
  it('itemBoxToRoom B', () => { expect(itemBoxToRoom('zz')).toMatchInlineSnapshot(`undefined`) })
  it('itemBoxToRoom C', () => { expect(itemBoxToRoom('C (couteau)')).toMatchInlineSnapshot(`"salon"`) })
  it('itemBoxToRoom D empty box', () => { expect(itemBoxToRoom('')).toMatchInlineSnapshot(`undefined`) })

  it('statusStringToStatus A lost', () => { expect(statusStringToStatus('lost')).toMatchInlineSnapshot(`"lost"`) })
  it('statusStringToStatus B bought', () => { expect(statusStringToStatus('bought')).toMatchInlineSnapshot(`"bought"`) })
  it('statusStringToStatus C to-give', () => { expect(statusStringToStatus('to-give')).toMatchInlineSnapshot(`"to-give"`) })
  it('statusStringToStatus D for-sell', () => { expect(statusStringToStatus('for-sell')).toMatchInlineSnapshot(`"for-sell"`) })
  it('statusStringToStatus E un-handled status', () => { expect(statusStringToStatus('hehe')).toMatchInlineSnapshot(`"bought"`) })

  it('drawerStringToDrawer A valid', () => { expect(drawerStringToDrawer('2')).toBe(2) })
  it('drawerStringToDrawer B empty', () => { expect(drawerStringToDrawer('')).toBe(-1) })
  it('drawerStringToDrawer C NaN', () => { expect(drawerStringToDrawer('hehe')).toBe(-1) })

  it('boxStringToBox A valid', () => { expect(boxStringToBox('A (apple)')).toMatchInlineSnapshot(`"A (apple)"`) })
  it('boxStringToBox B invalid', () => { expect(boxStringToBox('hehe')).toBe("") })

  it('isDataOlderThan A', () => {
    expect(isDataOlderThan(0, 0), 'data is considered older if time 0').toBe(true)
  })

  it('isDataOlderThan B', async () => {
    const timestamp = Date.now()
    await sleep(100)
    expect(isDataOlderThan(50, timestamp), 'data is older than 50ms after waiting 100ms').toBe(true)
  })

  it('isDataOlderThan C', async () => {
    const timestamp = Date.now()
    await sleep(50)
    expect(isDataOlderThan(100, timestamp), 'data is not older than 100ms after waiting 50ms').toBe(false)
  })

  it('isDataOlderThan D 2 minutes older than 1 minute', () => {
    const twoMinutesAgo = Date.now() - 2 * 60 * 1000
    expect(isDataOlderThan(nbMsInMinute, twoMinutesAgo)).toBe(true)
  })

  it('getItems A items empty, items not fresh => fetch successful', async () => {
    const twoMinutesAgo = Date.now() - 2 * 60 * 1000
    const result = await getItems([], twoMinutesAgo)
    const { error, value } = Result.unwrap(result)
    expect(error).toBe(undefined)
    expect(value?.includes('0 items loaded')).toBe(true)
    expect(databaseMock.listDocuments).toHaveBeenNthCalledWith(1, '', '', [{ isThisMockedDataFromMock: true, limit: 100 }, { isThisMockedDataFromMock: true, offset: 0 }])
  })

  it('getItems B items empty, items fresh => fetch successful', async () => {
    const twoMsAgo = Date.now() - 2
    const result = await getItems([], twoMsAgo)
    const { error, value } = Result.unwrap(result)
    expect(error).toBe(undefined)
    expect(value?.includes('0 items loaded')).toBe(true)
    expect(databaseMock.listDocuments).toHaveBeenNthCalledWith(1, '', '', [{ isThisMockedDataFromMock: true, limit: 100 }, { isThisMockedDataFromMock: true, offset: 0 }])
  })

  it('getItems C items not empty, items not fresh => fetch successful', async () => {
    const twoMinutesAgo = Date.now() - 2 * 60 * 1000
    const result = await getItems([mockItem()], twoMinutesAgo)
    const { error, value } = Result.unwrap(result)
    expect(error).toBe(undefined)
    expect(value?.includes('0 items loaded')).toBe(true)
    expect(databaseMock.listDocuments).toHaveBeenNthCalledWith(1, '', '', [{ isThisMockedDataFromMock: true, limit: 100 }, { isThisMockedDataFromMock: true, offset: 0 }])
  })

  it('getItems D items not empty, items not fresh => fetch failed', async () => {
    databaseMock.listDocuments.mockRejectedValueOnce(new Error('some error'))
    const twoMinutesAgo = Date.now() - 2 * 60 * 1000
    const result = await getItems([mockItem()], twoMinutesAgo)
    const { error } = Result.unwrap(result)
    expect(error).toMatchInlineSnapshot(`[Error: some error]`)
    expect(databaseMock.listDocuments).toHaveBeenNthCalledWith(1, '', '', [{ isThisMockedDataFromMock: true, limit: 100 }, { isThisMockedDataFromMock: true, offset: 0 }])
  })

  it('getItems E items not empty, items fresh => no fetch', async () => {
    const twoMsAgo = Date.now() - 2
    const result = await getItems([mockItem()], twoMsAgo)
    const { error, value } = Result.unwrap(result)
    expect(error).toBe(undefined)
    expect(value).toMatchInlineSnapshot(`"tasks are fresh (now)"`)
    expect(databaseMock.listDocuments).not.toHaveBeenCalled()
  })

})
