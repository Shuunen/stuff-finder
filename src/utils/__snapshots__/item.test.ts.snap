// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`deleteItem A delete an item in state 1`] = `
{
  "issues": undefined,
  "output": {
    "deleted": true,
    "id": "itemA",
  },
  "success": true,
  "typed": true,
}
`;

exports[`formToItem A default itemForm 1`] = `
{
  "barcode": "",
  "box": "",
  "brand": "",
  "category": "",
  "details": "",
  "drawer": "",
  "id": "",
  "location": "",
  "name": "",
  "photo": [],
  "price": undefined,
  "ref-printed": false,
  "reference": "",
  "status": "acheté",
  "updated-on": "",
}
`;

exports[`formToItem B itemForm with some values 1`] = `
{
  "barcode": "",
  "box": "",
  "brand": "",
  "category": "",
  "details": "",
  "drawer": "",
  "id": "",
  "location": "",
  "name": "",
  "photo": [
    {
      "url": "https://photos.org/200/200",
    },
  ],
  "price": 42,
  "ref-printed": false,
  "reference": "",
  "status": "acheté",
  "updated-on": "",
}
`;

exports[`getCommonListsFromItems A 1`] = `
{
  "boxes": [
    "",
    "G (brico & sport)",
  ],
  "categories": [
    "",
    "category B",
  ],
  "drawers": [
    "",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
  ],
  "locations": [
    "",
    "location A",
    "Salon",
  ],
  "statuses": [
    "",
    "à donner",
    "à vendre",
    "acheté",
    "donné",
    "jeté",
    "new surprise status",
    "renvoyé",
    "vendu",
  ],
}
`;

exports[`getOneItem A success result 1`] = `
{
  "barcode": "",
  "box": "box A",
  "brand": "",
  "category": "category A",
  "details": "",
  "drawer": "",
  "id": "rec123",
  "location": "location A",
  "name": "item A",
  "photo": [],
  "price": 0,
  "ref-printed": false,
  "reference": "",
  "status": "acheté",
  "updated-on": "",
}
`;

exports[`itemToForm A mocked item 1`] = `
{
  "columns": 10,
  "errorMessage": "",
  "fields": {
    "barcode": {
      "columns": 4,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Barcode",
      "link": "",
      "regex": /\\^\\[-,\\\\d\\\\p\\{L\\}\\\\s/&\\]\\{3,30\\}\\$/u,
      "type": "text",
      "unit": "",
      "value": "barcode B",
    },
    "box": {
      "columns": 4,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Box",
      "link": "",
      "options": [],
      "regex": /\\^\\[\\\\p\\{L\\}\\\\s&\\(\\)\\]\\{3,100\\}\\$/u,
      "type": "select",
      "unit": "",
      "value": "G (brico & sport)",
    },
    "brand": {
      "columns": 2,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Brand",
      "link": "",
      "regex": /\\^\\[-,\\\\d\\\\p\\{L\\}\\\\s/&\\]\\{3,100\\}\\$/u,
      "type": "text",
      "unit": "",
      "value": "brand B",
    },
    "category": {
      "columns": 3,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Category",
      "link": "",
      "options": [],
      "regex": /\\^\\[-,\\\\d\\\\p\\{L\\}\\\\s/&\\]\\{3,100\\}\\$/u,
      "type": "select",
      "unit": "",
      "value": "category B",
    },
    "details": {
      "columns": 4,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Details",
      "link": "",
      "regex": /\\^\\[-,\\\\d\\\\p\\{L\\}\\\\s/&\\]\\{3,200\\}\\$/u,
      "type": "text",
      "unit": "",
      "value": "details B",
    },
    "drawer": {
      "columns": 3,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Drawer",
      "link": "",
      "options": [
        {
          "label": "",
          "value": "",
        },
        {
          "label": "1",
          "value": "1",
        },
        {
          "label": "2",
          "value": "2",
        },
        {
          "label": "3",
          "value": "3",
        },
        {
          "label": "4",
          "value": "4",
        },
        {
          "label": "5",
          "value": "5",
        },
        {
          "label": "6",
          "value": "6",
        },
        {
          "label": "7",
          "value": "7",
        },
      ],
      "regex": /\\\\d/u,
      "type": "select",
      "unit": "",
      "value": "2",
    },
    "isPrinted": {
      "columns": 1,
      "isRequired": false,
      "isValid": false,
      "isVisible": false,
      "label": "Printed",
      "link": "",
      "regex": /\\^\\[-,\\\\d\\\\p\\{L\\}\\\\s/&\\]\\{3,100\\}\\$/u,
      "type": "checkbox",
      "unit": "",
      "value": false,
    },
    "location": {
      "columns": 3,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Location",
      "link": "",
      "options": [],
      "regex": /\\^\\[-,\\\\d\\\\p\\{L\\}\\\\s/&\\]\\{3,100\\}\\$/u,
      "type": "select",
      "unit": "",
      "value": "Salon",
    },
    "name": {
      "columns": 4,
      "isRequired": true,
      "isValid": false,
      "isVisible": true,
      "label": "Name",
      "link": "",
      "regex": /\\^\\[-,\\\\d\\\\p\\{L\\}\\\\s/&\\]\\{3,100\\}\\$/u,
      "type": "text",
      "unit": "",
      "value": "name B",
    },
    "photo": {
      "columns": 4,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Photo",
      "link": "",
      "regex": /\\^https\\?:\\\\/\\\\/\\\\S\\+\\$/u,
      "type": "text",
      "unit": "",
      "value": "https://picsum.photos/seed/unknown-name/200/200",
    },
    "price": {
      "columns": 3,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Price",
      "link": "",
      "regex": /\\^\\\\d\\{1,5\\}\\$/u,
      "type": "text",
      "unit": "€",
      "value": "undefined",
    },
    "reference": {
      "columns": 3,
      "isRequired": true,
      "isValid": false,
      "isVisible": true,
      "label": "Reference",
      "link": "",
      "regex": /\\^\\[\\\\w-\\]\\{3,50\\}\\$/u,
      "type": "text",
      "unit": "",
      "value": "reference B",
    },
    "status": {
      "columns": 3,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Status",
      "link": "",
      "options": [
        {
          "label": "acheté",
          "value": "acheté",
        },
        {
          "label": "à donner",
          "value": "à donner",
        },
        {
          "label": "à vendre",
          "value": "à vendre",
        },
        {
          "label": "donné",
          "value": "donné",
        },
        {
          "label": "jeté",
          "value": "jeté",
        },
        {
          "label": "renvoyé",
          "value": "renvoyé",
        },
        {
          "label": "vendu",
          "value": "vendu",
        },
      ],
      "regex": /\\^\\[-,\\\\d\\\\p\\{L\\}\\\\s/&\\]\\{3,100\\}\\$/u,
      "type": "select",
      "unit": "",
      "value": "acheté",
    },
  },
  "isTouched": false,
  "isValid": false,
}
`;

exports[`itemToForm C no item 1`] = `
{
  "columns": 10,
  "errorMessage": "",
  "fields": {
    "barcode": {
      "columns": 4,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Barcode",
      "link": "",
      "regex": /\\^\\[-,\\\\d\\\\p\\{L\\}\\\\s/&\\]\\{3,30\\}\\$/u,
      "type": "text",
      "unit": "",
      "value": "",
    },
    "box": {
      "columns": 4,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Box",
      "link": "",
      "options": [],
      "regex": /\\^\\[\\\\p\\{L\\}\\\\s&\\(\\)\\]\\{3,100\\}\\$/u,
      "type": "select",
      "unit": "",
      "value": "",
    },
    "brand": {
      "columns": 2,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Brand",
      "link": "",
      "regex": /\\^\\[-,\\\\d\\\\p\\{L\\}\\\\s/&\\]\\{3,100\\}\\$/u,
      "type": "text",
      "unit": "",
      "value": "",
    },
    "category": {
      "columns": 3,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Category",
      "link": "",
      "options": [],
      "regex": /\\^\\[-,\\\\d\\\\p\\{L\\}\\\\s/&\\]\\{3,100\\}\\$/u,
      "type": "select",
      "unit": "",
      "value": "",
    },
    "details": {
      "columns": 4,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Details",
      "link": "",
      "regex": /\\^\\[-,\\\\d\\\\p\\{L\\}\\\\s/&\\]\\{3,200\\}\\$/u,
      "type": "text",
      "unit": "",
      "value": "",
    },
    "drawer": {
      "columns": 3,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Drawer",
      "link": "",
      "options": [
        {
          "label": "",
          "value": "",
        },
        {
          "label": "1",
          "value": "1",
        },
        {
          "label": "2",
          "value": "2",
        },
        {
          "label": "3",
          "value": "3",
        },
        {
          "label": "4",
          "value": "4",
        },
        {
          "label": "5",
          "value": "5",
        },
        {
          "label": "6",
          "value": "6",
        },
        {
          "label": "7",
          "value": "7",
        },
      ],
      "regex": /\\\\d/u,
      "type": "select",
      "unit": "",
      "value": "",
    },
    "isPrinted": {
      "columns": 1,
      "isRequired": false,
      "isValid": false,
      "isVisible": false,
      "label": "Printed",
      "link": "",
      "regex": /\\^\\[-,\\\\d\\\\p\\{L\\}\\\\s/&\\]\\{3,100\\}\\$/u,
      "type": "checkbox",
      "unit": "",
      "value": false,
    },
    "location": {
      "columns": 3,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Location",
      "link": "",
      "options": [],
      "regex": /\\^\\[-,\\\\d\\\\p\\{L\\}\\\\s/&\\]\\{3,100\\}\\$/u,
      "type": "select",
      "unit": "",
      "value": "",
    },
    "name": {
      "columns": 4,
      "isRequired": true,
      "isValid": false,
      "isVisible": true,
      "label": "Name",
      "link": "",
      "regex": /\\^\\[-,\\\\d\\\\p\\{L\\}\\\\s/&\\]\\{3,100\\}\\$/u,
      "type": "text",
      "unit": "",
      "value": "",
    },
    "photo": {
      "columns": 4,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Photo",
      "link": "",
      "regex": /\\^https\\?:\\\\/\\\\/\\\\S\\+\\$/u,
      "type": "text",
      "unit": "",
      "value": "",
    },
    "price": {
      "columns": 3,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Price",
      "link": "",
      "regex": /\\^\\\\d\\{1,5\\}\\$/u,
      "type": "text",
      "unit": "€",
      "value": "",
    },
    "reference": {
      "columns": 3,
      "isRequired": true,
      "isValid": false,
      "isVisible": true,
      "label": "Reference",
      "link": "",
      "regex": /\\^\\[\\\\w-\\]\\{3,50\\}\\$/u,
      "type": "text",
      "unit": "",
      "value": "",
    },
    "status": {
      "columns": 3,
      "isRequired": false,
      "isValid": false,
      "isVisible": true,
      "label": "Status",
      "link": "",
      "options": [
        {
          "label": "acheté",
          "value": "acheté",
        },
        {
          "label": "à donner",
          "value": "à donner",
        },
        {
          "label": "à vendre",
          "value": "à vendre",
        },
        {
          "label": "donné",
          "value": "donné",
        },
        {
          "label": "jeté",
          "value": "jeté",
        },
        {
          "label": "renvoyé",
          "value": "renvoyé",
        },
        {
          "label": "vendu",
          "value": "vendu",
        },
      ],
      "regex": /\\^\\[-,\\\\d\\\\p\\{L\\}\\\\s/&\\]\\{3,100\\}\\$/u,
      "type": "select",
      "unit": "",
      "value": "acheté",
    },
  },
  "isTouched": false,
  "isValid": false,
}
`;

exports[`pushItem A item with no id should post remotely & push locally 1`] = `
{
  "issues": undefined,
  "output": {
    "createdTime": "",
    "fields": {
      "barcode": "",
      "box": "box A",
      "brand": "",
      "category": "category A",
      "details": "",
      "drawer": "",
      "location": "location A",
      "name": "item A",
      "ref-printed": false,
      "reference": "reference-a",
      "status": "acheté",
      "updated-on": "2022-12-26T15:42:00.000Z",
    },
    "id": "freshly-created",
  },
  "success": true,
  "typed": true,
}
`;

exports[`pushItem A item with no id should post remotely & push locally 3`] = `
{
  "fields": {
    "barcode": "barcode B",
    "box": "G (brico & sport)",
    "brand": "brand B",
    "category": "category B",
    "details": "details B",
    "drawer": "2",
    "location": "Salon",
    "name": "name B",
    "photo": [
      {
        "url": "https://picsum.photos/seed/unknown-name/200/200",
      },
    ],
    "ref-printed": false,
    "reference": "reference B",
    "status": "acheté",
  },
}
`;

exports[`pushItem B item with id should patch remotely & update locally 1`] = `
{
  "issues": undefined,
  "output": {
    "createdTime": "",
    "fields": {
      "barcode": "",
      "box": "box A",
      "brand": "",
      "category": "category A",
      "details": "",
      "drawer": "",
      "location": "location A",
      "name": "item A",
      "ref-printed": false,
      "reference": "reference-a",
      "status": "acheté",
      "updated-on": "2022-12-26T15:42:00.000Z",
    },
    "id": "existing",
  },
  "success": true,
  "typed": true,
}
`;

exports[`pushItem B item with id should patch remotely & update locally 3`] = `
{
  "fields": {
    "details": "new details",
  },
}
`;
